### PR TITLE
Nicos/navmesh

### PR DESCRIPTION
**Pull request by nicolasgustafsson**
Caution: It's not just a navmesh

### CODE REVIEW

File: Game/Designers/LevelDesigner/Geometry/Polygon.cpp [hunk 1]
```diff
@@ -1,10 +1,31 @@
 #include "pch.h"
 #include "Polygon.h"
 #include "Randomizer/Randomizer.h"
+#include <Utilities/Rotation.h>
+#include "Physics/Navmesh/Navmesh.h"
 
-namespace Geometry
+float sign(SVector2f p1, SVector2f p2, SVector2f p3)
```
Comment (Devilhar) [inline Game/Designers/LevelDesigner/Geometry/Polygon.cpp:7]: Use f32 and rename to Sign

File: Game/Designers/LevelDesigner/Geometry/Polygon.cpp [hunk 2]
```diff
@@ -1,10 +1,31 @@
...
+{
+	return (p1.X - p3.X) * (p2.Y - p3.Y) - (p2.X - p3.X) * (p1.Y - p3.Y);
+}
+
+bool PointInTriangle(SVector2f pt, SVector2f v1, SVector2f v2, SVector2f v3)
 {
+	float d1, d2, d3;
```
Comment (Devilhar) [inline Game/Designers/LevelDesigner/Geometry/Polygon.cpp:14]: f32

File: Game/Designers/LevelDesigner/Geometry/Polygon.cpp [hunk 3]
```diff
@@ -56,4 +77,158 @@ namespace Geometry
...
+		// equations-to-code conversion
+		const f32 x1 = p1.X, x2 = p2.X, x3 = p3.X, x4 = p4.X;
+		const f32 y1 = p1.Y, y2 = p2.Y, y3 = p3.Y, y4 = p4.Y;
+
+		// calculate the distance to intersection point
+		const f32 uA = ((x4 - x3) * (y1 - y3) - (y4 - y3) * (x1 - x3)) / ((y4 - y3) * (x2 - x1) - (x4 - x3) * (y2 - y1));
+		const f32 uB = ((x2 - x1) * (y1 - y3) - (y2 - y1) * (x1 - x3)) / ((y4 - y3) * (x2 - x1) - (x4 - x3) * (y2 - y1));
+
+		// if uA and uB are between 0-1, lines are colliding
+		if (uA >= aTolerance && uA <= 1.f - aTolerance && uB >= aTolerance && uB <= 1.f - aTolerance) {
```
Comment (Devilhar) [inline Game/Designers/LevelDesigner/Geometry/Polygon.cpp:219]: Put { on a new line

File: Game/Physics/Navmesh/Navmesh.cpp [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+	{
+		SLine line{ aPosition, *point };
+		if (myPolygon.IntersectsWithAnySide(line, point))
+		{
+			++point;
+			continue;
+		}
+
+		//if ((*point).DistanceTo({ 730.f, 25.f }) < 1.f)
+		//	WmLog("");
```
Comment (Devilhar) [inline Game/Physics/Navmesh/Navmesh.cpp:36]: Remove this

File: Game/Physics/Navmesh/Navmesh.cpp [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+			if (!aTriangle->FirstSide)
+				aTriangle->FirstSide = &(*foundLine);
+			else if (!aTriangle->SecondSide)
+				aTriangle->SecondSide = &(*foundLine);
+			else if (!aTriangle->ThirdSide)
+				aTriangle->ThirdSide = &(*foundLine);
+		}
+		else
+		{
+			WmLog("wee");
```
Comment (Devilhar) [inline Game/Physics/Navmesh/Navmesh.cpp:172]: Write more explicit comment, or handle adding the same line twice

File: Game/Physics/Navmesh/Navmesh.h [hunk 6]
```diff
@@ -0,0 +1,50 @@
+#pragma once
+#include "Designers/LevelDesigner/Geometry/PolygonOperators/PolygonPointOperator.h"
+#include <Designers/LevelDesigner/Geometry/Polygon.h>
```
Comment (Devilhar) [inline Game/Physics/Navmesh/Navmesh.h:3]: Change to:
#include "Designers/LevelDesigner/Geometry/Polygon.h"

File: Game/Utility/Palette.h [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+	const static SColor PrimaryColor = SColor::Amaranth;
+	const static SColor SecondaryColor = SColor::SeaGreenCrayola;
+	const static SColor AccentColor = SColor::MiddleYellow;
+	const static SColor BlackBackground = SColor::EerieBlack;
+	const static SColor DarkBackground = SColor::Jet;
+	const static SColor MediumBackground = SColor::SpanishGray;
+	const static SColor LightBackground = SColor::Gainsboro;
+
+	const static SColor EnemyColor = PrimaryColor;
+	const static SColor PlayerColor = SecondaryColor;
```
Comment (Devilhar) [inline Game/Utility/Palette.h:14]: Make constexpr

File: Game/UtilityFunctionalities/LineOfSightFunctionality.cpp [hunk 8]
```diff
@@ -0,0 +1,28 @@
+#include "pch.h"
+#include "LineOfSightFunctionality.h"
+#include "Levels/LevelDenizenFunctionality.h"
+#include <Levels/LevelFunctionality.h>
```
Comment (Devilhar) [inline Game/UtilityFunctionalities/LineOfSightFunctionality.cpp:4]: Change to:
#include "Levels/LevelFunctionality.h"

File: WonderMake/Camera/Camera.cpp [hunk 9]
```diff
@@ -3,8 +3,9 @@
 #include "Graphics/EngineUniformBuffer.h"
 #include "Program/Window.h"
 #include <GLFW/glfw3.h>
-
-REGISTER_SYSTEM(Camera);
+#include <Program/GlfwFacade.h>
```
Comment (Devilhar) [inline WonderMake/Camera/Camera.cpp:6]: Change to:
#include "Program/GlfwFacade.h"

File: WonderMake/Camera/Display.cpp [hunk 10]
```diff
@@ -0,0 +1,152 @@
+#include "pch.h"
+#include "Display.h"
+#include "Graphics/EngineUniformBuffer.h"
+#include "Program/Window.h"
+#include <GLFW/glfw3.h>
+#include <Program/GlfwFacade.h>
```
Comment (Devilhar) [inline WonderMake/Camera/Display.cpp:6]: Change to:
#include "Program/GlfwFacade.h"

File: WonderMake/Graphics/RenderNodeGraph/RenderNodeTypes.cpp [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+namespace NodeTypes
+{
+	void SRenderResultNode::ExecuteNode(SNode& aNode)
+	{
+		std::any& finalRenderTarget = aNode.NodeData["RenderTarget"];
+
+		finalRenderTarget = std::make_any<std::shared_ptr<RenderTarget>>(aNode.GetInput<std::shared_ptr<RenderTarget>>(0));
+	}
+
+	void SRenderTextureNode::ExecuteNode(struct SNode& aNode)
```
Comment (Devilhar) [inline WonderMake/Graphics/RenderNodeGraph/RenderNodeTypes.cpp:20]: Is struct required? If not, remove from all functions.

File: WonderMake/Graphics/RenderNodeGraph/RenderNodeTypes.h [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+			AddSlot<std::shared_ptr<RenderTarget>>(ESlotIo::Output, "Out rendertarget");
+		}
+
+		virtual void ExecuteNode(struct SNode&) override;
+	};
+}
+
+namespace SlotColors
+{
+}
```
Comment (Devilhar) [inline WonderMake/Graphics/RenderNodeGraph/RenderNodeTypes.h:84]: ?

File: WonderMake/Graphics/RenderObject.h [hunk 13]
```diff
@@ -21,32 +22,109 @@ struct SRenderObjectInfo
...
+
+	void SetRenderLayer(const std::string& aRenderLayer) { myRenderLayer = aRenderLayer; };
+	void SetRenderOrder(i32 aRenderOrder) noexcept { myRenderOrder = aRenderOrder; }
+protected:
+
+	virtual void RenderInternal() = 0;
+	i32 myRenderOrder = 0;
+
+	//[Nicos]: Can this be string_view?
+	std::string myRenderLayer = "Default";
```
Comment (Devilhar) [inline WonderMake/Graphics/RenderObject.h:42]: I don't see why not.

File: WonderMake/Graphics/RenderSettingsManager.cpp [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+	switch(aBlendMode)
+	{
+	case EBlendMode::Alpha:
+		SystemPtr<OpenGLFacade>()->SetBlendFunction(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
+		break;
+	case EBlendMode::Additive:
+		SystemPtr<OpenGLFacade>()->SetBlendFunction(GL_SRC_ALPHA, GL_ONE);
+		break;
+	default:
+		break;
```
Comment (Devilhar) [inline WonderMake/Graphics/RenderSettingsManager.cpp:63]: Remove this. We should be able to enable warnings whenever a case is missing from a switch statement.

File: WonderMake/NodeGraph/NodeGraph.cpp [hunk 15]
```diff
@@ -72,36 +79,24 @@ void NodeGraph::Load()
...
+		Compile();
 
-	for (size_t i = 0; i < myCompiledNodeStack.size(); i++)
+	for (size_t i = myCompiledNodeStack.size() - 1; i < myCompiledNodeStack.size(); i--)
 	{
 		auto&& compiledNode = myCompiledNodeStack[i];
 
-		compiledNode.Node.NodeType.ExecuteNodeLeftToRight(compiledNode.Node);
+		compiledNode.Node.NodeType.ExecuteNode(compiledNode.Node);
 	}
```
Comment (Devilhar) [inline WonderMake/NodeGraph/NodeGraph.cpp:100]: Why is this being done in reverse? The direction for unordered containers should not matter.

File: WonderMake/Threads/Routine.cpp [hunk 16]
```diff
@@ -19,15 +19,20 @@ void Routine::Run()
 
 void Routine::AddProcedure(Closure aClosure)
 {
-	myProcedures.push_back(aClosure);
+	myProceduresToAdd.insert(aClosure);
```
Comment (Devilhar) [inline WonderMake/Threads/Routine.cpp:22]: emplace and std::move

File: WonderMake/Utilities/plf_colony.h [hunk 17]
```diff
@@ -259,7 +259,7 @@ namespace plf
 		typedef typename element_allocator_type::size_type			size_type;
 		typedef typename element_allocator_type::difference_type	difference_type;
 		typedef typename element_allocator_type::reference			reference;
-		typedef typename element_allocator_type::const_reference	const_reference;
+		typedef typename element_allocator_type::const_reference	const_reference;
```
Comment (Devilhar) [inline WonderMake/Utilities/plf_colony.h:262]: ?

File: Bin/Shaders/Fragment/Test.frag [hunk 18]
```diff
@@ -0,0 +1,15 @@
+#version 440
```
[the remark this answers is not in the document]
Comment (nicolasgustafsson) [reply Bin/Shaders/Fragment/Test.frag]: Done

File: Bin/Shaders/Vertex/LineList.vert [hunk 19]
```diff
@@ -0,0 +1,16 @@
+#version 440
+layout (location = 0) in vec2 aPos;
+
+
+//!#include "Common.glsl"
+
+void main() 
+{
+  vec3 transformedPosition = ViewProjectionMatrix * vec3(aPos.x, aPos.y, 1.0);
```
[the remark this answers is not in the document]
Comment (nicolasgustafsson) [reply Bin/Shaders/Vertex/LineList.vert]: Done

File: Bin/Shaders/Vertex/NavmeshVertexShader.vert [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+layout (location = 2) in float aDepth;
+
+out vec4 vertexColor;
+out vec2 worldPos;
+
+//!#include "Common.glsl"
+
+void main() 
+{
+  vec3 transformedPosition = ViewProjectionMatrix * vec3(aPos.x, aPos.y, 1.0);
```
[the remark this answers is not in the document]
Comment (nicolasgustafsson) [reply Bin/Shaders/Vertex/NavmeshVertexShader.vert]: Done

**Pull request merged**